core/src/stage.cpp

```diff
@@ -217,9 +222,17 @@ void StagePrivate::newSolution(const SolutionBasePtr& solution) {
 }
 
 bool StagePrivate::addCost(SolutionBase& solution) {
+	double cost{ 0.0 };
+
 	auto* trajectory{ dynamic_cast<const SubTrajectory*>(&solution) };
 	if (trajectory && cost_term_)
-		solution.setCost(cost_term_(*trajectory));
+		// CostTerm only applies to SubTrajectory
+		cost = cost_term_(*trajectory);
+	else
+		// everything else is just transformed
+		cost = solution.cost();
+
+	solution.setCost(cost_transform_(cost));
```

Comment:
You can already do that with lambdas, because CostTerm is just a function signature:
```
Clearance clear;
stage.setCostTerm([&](auto&& s, auto&& comment){
   string a, b;
   double pl = PathLength(s,a);
   double cl = clear(s,b);
   comment = a + ", " + b;
   return c1*pl + c2*log(cl);
});
```
What do you hope to gain from "wrapper classes"?
You are right though in that the CostTransform concept can always be integrated with a WrapperStage modifying the cost.